src/ngx_http_lua_kong_log.c

```diff
@@ -14,45 +14,133 @@
+
+void
+ngx_http_lua_kong_log_set_ready_for_dynamic_log_level(ngx_int_t is_ready)
 {
-    ngx_log_t                   *log;
-    ngx_listening_t             *ls;
-    ngx_http_core_loc_conf_t    *clcf;
-    ngx_uint_t                   i;
+    /*
+    ** This function will be called when the following functions
+    ** are about to start event loop:
+    **   * ngx_worker_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_single_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_cache_manager_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_cache_loader_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **   * ngx_privileged_agent_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+    **
+    ** This function will be called immediately when the event loop has just been stopped.
+    **
+    ** Once the event loop is started,
+    ** we can confirm the data structure and other resources are ready for
+    ** the dynamic log level feature.
+    **
+    ** Please see the function ngx_worker_process_cycle in the file
+    ** src/os/unix/ngx_process_cycle.c.
+    */
+    g_ready_for_dynamic_log_level = is_ready;
+}
 
-    if (r == NULL) {
-        return NGX_ERROR;
-    }
+int32_t
+ngx_http_lua_kong_ffi_set_dynamic_log_level(uint32_t log_level, uint32_t timeout)
+{
+    ngx_http_lua_kong_main_conf_t *lkmcf =
+        ngx_http_cycle_get_module_main_conf(ngx_cycle, ngx_http_lua_kong_module);
 
-    clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
-    clcf->error_log->log_level = (ngx_uint_t) level;
+    ngx_http_lua_kong_assert(lkmcf != NULL);
+
+    if (timeout == 0) {
+        /*
+        ** if the timeout is 0,
+        ** we will disable the dynamic log level,
+        ** and ignore the first parameter.
+        */
+        lkmcf->dynmic_log_level_enabled = 0;
+        return NGX_OK;
+    }
 
-    if (ngx_cycle == NULL) {
-        return NGX_ERROR;
+    /*
+    ** if the log level is invalid,
+    ** we will not change the current log level,
+    ** and return an error.
+    */
+    switch (log_level)
+    {
+        case NGX_LOG_STDERR: case NGX_LOG_EMERG:
+        case NGX_LOG_ALERT:  case NGX_LOG_CRIT:
+        case NGX_LOG_ERR:    case NGX_LOG_WARN:
+        case NGX_LOG_NOTICE: case NGX_LOG_INFO:
+        case NGX_LOG_DEBUG:
+            break;
+        default:
+            return NGX_ERROR;
     }
 
-    if (r->connection && r->connection->log) {
-        log = r->connection->log;
+    lkmcf->dynmic_log_level = log_level;
+    lkmcf->dynmic_log_level_timeout_at = ngx_time() + timeout;
+    lkmcf->dynmic_log_level_enabled = 1;
 
-    } else {
-        log = ngx_cycle->log;
+    return NGX_OK;
+}
+
+ngx_uint_t
+ngx_http_lua_kong_get_dynamic_log_level(ngx_uint_t current_log_level)
+{
+    if (!g_ready_for_dynamic_log_level) {
+        /* this function might be called before the neccessary resources are ready */
```

Comment:
what are the resources? adding an example would be nice

Reply:
done